src/classes/PS_IntegrationService.cls

```diff
@@ -54,33 +54,143 @@ global class PS_IntegrationService {
+            new Map<String, Payment_Services_Configuration__c>();
 
         for (String service : configParams.keySet()) {
             for (Map<String, String> valueMap : configParams.get(service)) {
-                configurations.add(createConfiguration(service, valueMap));
+                Payment_Services_Configuration__c configuration = createConfiguration(service, valueMap);
+                configurationsToUpsert.put(configuration.Service_Key__c, configuration);
             }
         }
 
+        return configurationsToUpsert;
+    }
+
+    /*******************************************************************************************************
+    * @description Queries for existing Payment_Services_Configuration__c records
+    *
+    * @param serviceKeys A set of Payment_Services_Configuration__c.Service_Key__c values
+    */
+    private static Payment_Services_Configuration__c[] getExistingRecords(Set<String> serviceKeys) {
+        return [
+            SELECT Id,
+                Service_Key__c,
+                Key__c,
+                Value__c
+            FROM Payment_Services_Configuration__c
+            WHERE Service_Key__c = :serviceKeys
+        ];
+    }
+
+    /*******************************************************************************************************
+    * @description Applies the existing configuration record ids to the corresponding newly generated
+    * configuration in preparation for an update.
+    *
+    * @param existingConfigurations List of existing Payment_Services_Configuration__c records
+    * @param configurationsToUpsertByServiceKey Map of newly generated Payment_Services_Configuration__c records
+    * from the parameters provided by the caller of 'setConfiguration'.
+    */
+    private static void applyExistingIdsToNewConfigurations(
+        Payment_Services_Configuration__c[] existingConfigurations,
+        Map<String, Payment_Services_Configuration__c> configurationsToUpsertByServiceKey) {
+
+        for (Payment_Services_Configuration__c configuration : existingConfigurations) {
+            Payment_Services_Configuration__c newConfiguration =
+            configurationsToUpsertByServiceKey.get(configuration.Service_Key__c);
+            newConfiguration.Id = configuration.Id;
+        }
+    }
+
+    /*******************************************************************************************************
+    * @description Collects a set of Service_Key__c fields from a list of Payment_Services_Configuration__c
+    * records
+    *
+    * @param existingConfigurations List of existing Payment_Services_Configuration__c records
+    */
+    private static Set<String> collectServiceKeys(Payment_Services_Configuration__c[] existingConfigurations) {
+        Set<String> existingConfigurationKeys = new Set<String> {};
+
+        for (Payment_Services_Configuration__c configuration : existingConfigurations) {
+            existingConfigurationKeys.add(configuration.Service_Key__c);
```

Comment:
Sets allow `null` to be added, should we check for null before we add the value to the Set?